### PR TITLE
pkg/trace/api: distinguish timeout and EOF errors from decoding errors

### PR DESCRIPTION
**Pull request by richardstartin**
### What does this PR do?

Currently timeouts and unexpected EOF (which either indicate a real problem with the payload/Content-Length combination, or are caused by the client aborting the connection) are recorded as decoding errors, but often correspond to valid payloads which could have been decoded had the trace agent been given enough resources.

### Motivation

More accurate error metrics for diagnostic purposes

### Describe your test plan

I added a case where EOF is produced. I don't know how to simulate a connection timeout in Go but was hoping someone in the agent team could add this test case.

### CODE REVIEW

File: pkg/trace/api/api.go [hunk 1]
```diff
@@ -371,6 +371,10 @@ func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.
 		httpDecodingError(err, []string{"handler:traces", fmt.Sprintf("v:%s", v)}, w)
 		if err == ErrLimitedReaderLimitReached {
 			atomic.AddInt64(&ts.TracesDropped.PayloadTooLarge, tracen)
+		} else if err == io.EOF || err == io.ErrUnexpectedEOF {
+			atomic.AddInt64(&ts.TracesDropped.UnexpectedEOF, tracen)
+		} else if err, ok := err.(net.Error); ok && err.Timeout() {
+			atomic.AddInt64(&ts.TracesDropped.Timeout, tracen)
 		} else {
 			atomic.AddInt64(&ts.TracesDropped.DecodingError, tracen)
 		}
```
Comment (gbbr) [inline pkg/trace/api/api.go:380]: Let's use a switch-case here, it'll be nicer, I think
```suggestion
		switch err {
		case ErrLimitedReaderLimitReached:
			atomic.AddInt64(&ts.TracesDropped.PayloadTooLarge, tracen)
		case io.EOF, io.ErrUnexpectedEOF:
			atomic.AddInt64(&ts.TracesDropped.UnexpectedEOF, tracen)
		default:
		    if err, ok := err.(net.Error); ok && err.Timeout() {
			    atomic.AddInt64(&ts.TracesDropped.Timeout, tracen)
		    } else {
			    atomic.AddInt64(&ts.TracesDropped.DecodingError, tracen)
		    }
	    }
```

File: pkg/trace/api/api_test.go [hunk 2]
```diff
@@ -368,6 +368,31 @@ func TestReceiverDecodingError(t *testing.T) {
...
 }
 
+func TestReceiverUnexpectedEOF(t *testing.T) {
+	assert := assert.New(t)
+	conf := newTestReceiverConfig()
+	r := newTestReceiverFromConfig(conf)
+	server := httptest.NewServer(http.HandlerFunc(r.handleWithVersion(v05, r.handleTraces)))
+	// we get to read the header and the entire dictionary, but the request claims to be much larger
+	data := []byte{0x92, 0x91, 0xA5, 'a', 'b', 'c', 'd', 'e'}
+	client := &http.Client{}
```
Comment (gbbr) [inline pkg/trace/api/api_test.go:378]:
```suggestion
	var client http.Client
```
Your version is also correct, but this is the general convention.

File: pkg/trace/api/api_test.go [hunk 3]
```diff
@@ -368,6 +368,31 @@ func TestReceiverDecodingError(t *testing.T) {
 	})
 }
 
+func TestReceiverUnexpectedEOF(t *testing.T) {
+	assert := assert.New(t)
+	conf := newTestReceiverConfig()
+	r := newTestReceiverFromConfig(conf)
+	server := httptest.NewServer(http.HandlerFunc(r.handleWithVersion(v05, r.handleTraces)))
```
Comment (gbbr) [inline pkg/trace/api/api_test.go:375]: This is fine, but I'm curious - why not `r.Start` to use what actually runs?

File: pkg/trace/api/api_test.go [hunk 4]
```diff
@@ -368,6 +368,31 @@ func TestReceiverDecodingError(t *testing.T) {
...
+	// we get to read the header and the entire dictionary, but the request claims to be much larger
```
Comment (gbbr) [inline pkg/trace/api/api_test.go:376]: Not sure it's clear what the first 3 request bytes are meant to be from the description.

File: pkg/trace/api/api_test.go [hunk 5]
```diff
@@ -368,6 +368,31 @@ func TestReceiverDecodingError(t *testing.T) {
...
+	data := []byte{0x92, 0x91, 0xA5, 'a', 'b', 'c', 'd', 'e'}
+	client := &http.Client{}
+	traceCount := 2
+
+	t.Run("Content-Length larger than input", func(t *testing.T) {
```
Comment (gbbr) [inline pkg/trace/api/api_test.go:381]: You don't need a sub-test here, because we only have 1 test.

Also, I want to point out a common misconception about Go sub-tests. In most languages the quoted string argument represents a description of what the unit test does. For Go, it's generally just one word and it merely serves as an identifier to be able to trace back the test when you see it failing in your output, and not as a description. If you want to dig deeper, feel free to grep the standard library for `t.Run` or for `*testing.T` to see what the common names and conventions are. One hint that something is wrong is running the tests with `go test -v` and observing the inserted underscores in place of spaces.

In this case I'd just recommend using `unexpectedEOF` or something, but as I've said above it's not needed here.

File: pkg/trace/api/responses.go [hunk 6]
```diff
@@ -46,6 +47,17 @@ func httpDecodingError(err error, tags []string, w http.ResponseWriter) {
...
 		msg = errtag
 	}
+	// is it a timeout?
+	if err, ok := err.(net.Error); ok && err.Timeout() {
+		status = http.StatusRequestTimeout
+		errtag := "timeout"
+		msg = errtag
+	}
+	// is there something wrong with the content length?
+	if err == io.EOF || err == io.ErrUnexpectedEOF {
```
Comment (gbbr) [inline pkg/trace/api/responses.go:57]: Can we go with the similar switch-case here too?

File: pkg/trace/api/responses.go [hunk 7]
```diff
@@ -46,6 +47,17 @@ func httpDecodingError(err error, tags []string, w http.ResponseWriter) {
 		errtag := "payload-too-large"
 		msg = errtag
 	}
+	// is it a timeout?
```
Comment (gbbr) [inline pkg/trace/api/responses.go:50]: These inline comments can be removed. I don't think the code is complex enough to warrant clarification.

File: pkg/trace/info/stats.go [hunk 8]
```diff
@@ -204,6 +204,11 @@ type TracesDropped struct {
 	SpanIDZero int64
 	// ForeignSpan is when a span in a trace has a TraceId that is different than the first span in the trace
 	ForeignSpan int64
+	// Timeout is when a request times out
```
Comment (gbbr) [inline pkg/trace/info/stats.go:207]:
```suggestion
	// Timeout counts the number of payloads dropped due to read timeouts.
```

File: pkg/trace/info/stats.go [hunk 9]
```diff
@@ -204,6 +204,11 @@ type TracesDropped struct {
...
+	Timeout int64
+	// UnexpectedEOF is when an unexpected EOF is encountered, this can happen because the client has aborted
```
Comment (gbbr) [inline pkg/trace/info/stats.go:209]:
```suggestion
	// UnexpectedEOF is the number of times an unexpected EOF is encountered. This can happen because the client has aborted
```

**Review by gbbr — CHANGES_REQUESTED** [review]
Thanks for doing this. I was meaning to do this work as well.

Your implementation and all of it is fine, I just left comments regarding consistency and general Go idioms.

File: pkg/trace/api/responses.go [hunk 10]
```diff
@@ -46,6 +47,17 @@ func httpDecodingError(err error, tags []string, w http.ResponseWriter) {
...
+	if err, ok := err.(net.Error); ok && err.Timeout() {
+		status = http.StatusRequestTimeout
+		errtag := "timeout"
+		msg = errtag
+	}
+	// is there something wrong with the content length?
+	if err == io.EOF || err == io.ErrUnexpectedEOF {
+		errtag = "unexpected-eof"
```
Comment (gbbr) [inline pkg/trace/api/responses.go:58]: One last thing: since we're treating both EOF and UnexpectedEOF as the same, can we shorten this to just `eof`?

File: pkg/trace/api/api_test.go (hunk 2 above)
Comment (richardstartin) [reply pkg/trace/api/api_test.go:378]: Well, this isn't the convention used elsewhere in this file

File: pkg/trace/api/api_test.go (hunk 3 above)
Comment (richardstartin) [reply pkg/trace/api/api_test.go:375]: This was copied and pasted from elsewhere in this file, so I guess you could ask the same question about `TestReceiverDecodingError`

Screenshot 2020-09-07 at 13 16 08

File: pkg/trace/info/stats.go (hunk 8 above)
Comment (richardstartin) [reply pkg/trace/info/stats.go:207]: Sure, but it's using the same sentence structure as all the other ones?

File: pkg/trace/api/responses.go (hunk 10 above)
Comment (richardstartin) [reply pkg/trace/api/responses.go:58]: Yes perhaps

File: pkg/trace/info/stats.go (hunk 8 above)
Comment (gbbr) [reply pkg/trace/info/stats.go:207]: Yeah, I share the sentiment. Unfortunately that's the thing with this codebase. Some of it is older, some of it is newer. I'm trying to aim and make sure that at the very least the new code respects common idioms. Please consider this when seeing my comments, even if you copied code from elsewhere that might not. If you'd like, it's fine to change it everywhere but I wouldn't want to ask that of you.

File: pkg/trace/api/api_test.go (hunk 3 above)
Comment (gbbr) [reply pkg/trace/api/api_test.go:375]: 🤷‍♂️ Indeed. I think `r.Start` will be better because we're testing the actual thing that the agent runs, without leaving anything out. No idea why I did that then or what the history of that code is.

File: pkg/trace/api/api_test.go (hunk 2 above)
Comment (richardstartin) [reply pkg/trace/api/api_test.go:378]: This is the way you wrote it yourself elsewhere.
Screenshot 2020-09-07 at 13 22 43

Comment (gbbr) [reply pkg/trace/api/api_test.go:378]: Doesn't matter. This is what we want moving forward.

Comment (gbbr) [reply pkg/trace/api/api_test.go:378]: Richard, I didn't write the entire codebase, I copied it from its old location at https://github.com/DataDog/datadog-trace-agent so I show up on the entire code (minus recent changes since the copy). Can we focus on the actual work and moving the PR forward?

**Comment by richardstartin** [comment]
@gbbr I hoped this would help. If it was on your roadmap to fix this anyway, I'll leave it to you.

**Comment by gbbr** [comment]
It absolutely does help and I was hoping we could work together to merge it.

**Comment by gbbr** [comment]
Couldn't push here so re-opened https://github.com/DataDog/datadog-agent/pull/6329.
